keps/sig-cluster-lifecycle/draft-20180130-kubeadm-join-master.md

```diff
@@ -0,0 +1,430 @@
+
+  NB 2. Nothing in this proposal should prevent users to run master nodes components
+  and etcd on the same machines; however users should be aware that this will
+  introduce limitations for strategies like parallel node creation and In-place
+  vs. Replace reconciliation.
+
+  NB 3. Nothing in this proposal should prevent kubeadm to implement in future a
+  solution for provisioning an etcd cluster based on static pods/pods.
+
+* This proposal doesn't include a solution for API server load balancing.
+
+  NB. Nothing in this proposal should prevent users from choosing their preferred
+  solution for API server load balancing.
+
+* Nothing in this proposal should prevent practices that exist today.
+
+### Challenges and Open Questions
+
+* Keep the UX simple.
+
+  * _What are the acceptable trade-offs between the need to have a clean and simple
+    UX and the complexity of the following challenges and open questions?_
+
+* Create a cluster without knowing its final layout
+
+  Supporting a dynamic workflow implies that some information about the cluster are
+  not available at init time, like e.g. the number of master nodes, the ip of
+  master nodes etc. etc.
+
+  * _How to configure a Kubernetes cluster in order to easily adapt to future change
+    of its own layout like e.g. add a master node, remove a master node?_
+
+  * _What are the "pivotal" cluster settings that must be defined before initialising
+    the cluster?_
+
+  * _What are the mandatory conditions to be verified when executing`kubeadm init`
+    to allow/not allow the execution of `kubeadm join --master` in future?_
+
+* Kubeadm limited scope of action
+
+  * Kubeadm binary can execute actions _only_ on the machine where it is running
+    e.g. it is not possible to execute actions on other nodes, to copy files across
+    nodes etc.
+  * During the join workflow, kubeadm can access the cluster _only_ using identities
+    with limited grants, `system:unauthenticated` or `system:node-bootstrapper`.
+
+* Dependencies graduation
+
+  The solution for `kubeadm join --master` will rely on a set of dependencies/other
+  features which are still in the process for graduating to GA like e.g. dynamic
+  kubelet configuration, self-hosting, component config.
+
+  * _When `kubeadm join --master` should rely entirely on dependencies/features
+    still under graduation vs provide compatibility with older/less convenient but
+    more consolidated approaches?_
+
+  * _Should we support  `kubeadm join --master` for cluster with a control plane
+    deployed as static pods? What about cluster with a self-hosted
+    control plane?_
+  * _Should we support `kubeadm join --master` only  for cluster storing
+    cluster-certificates on file system? What about cluster
+    storing certificates in secrets?_
+
+* Upgradability
+
+  * How to setup an high available cluster in order to simplify the execution
+    of cluster updates, both manually or with the support of `kubeadm update`?_
```

Comment:
`cluster upgrades` (just to be consistent) and `kubeadm upgrade` (because `update` is not a command)